<launch> 
  <arg name="if_fix_final" value="false"/> 
  <arg name="start_x_" value="-4"/>
  <arg name="start_y_" value="-4"/>
  <arg name="start_yaw_" value="0"/>
  <arg name="final_x_" value="4"/>
  <arg name="final_y_" value="4"/>
  <arg name="final_yaw_" value="0"/>

  <!-- debug 走固定起点终点 -->  
  <!-- <arg name="if_fix_final" value="true"/>
  <arg name="start_x_" value="-1.81292"/>
  <arg name="start_y_" value="-1.80743"/>
  <arg name="start_yaw_" value="0.770891"/>
  <arg name="final_x_" value="3.11711"/>
  <arg name="final_y_" value="3.31731"/>
  <arg name="final_yaw_" value="0"/> -->



  <!-- 控制器 -->
  <arg name="if_mpc_" value="true"/>

  <node name="global_planning" pkg="plan_manager" type="global_planning" output="screen">
    <rosparam file="$(find front_end)/config/kinoastar.yaml" command="load"/>
    <rosparam file="$(find plan_manager)/config/car.yaml" command="load"/>
    <rosparam file="$(find back_end)/config/global_planning1ms.yaml" command="load"/>
    <param name="start_x" value="$(arg start_x_)"/>
    <param name="start_y" value="$(arg start_y_)"/>
    <param name="start_yaw" value="$(arg start_yaw_)"/>
    <param name="if_fix_final" value="$(arg if_fix_final)"/>
    <param name="final_x" value="$(arg final_x_)"/>
    <param name="final_y" value="$(arg final_y_)"/>
    <param name="final_yaw" value="$(arg final_yaw_)"/>
    <!-- 重规划 -->
    <param name="replan_time" value="0.5"/>
    <!-- <param name="replan_time" value="50"/> -->
    <!-- 规划器的最大运算时间，也可以认为是规划器的预测范围 -->
    <param name="max_replan_time" value="0.1"/>
    <!-- 控制器 -->
    <param name="if_mpc" value="$(arg if_mpc_)"/>
    <!-- 可探测范围 -->
    <param name="detection_range" value="10"/>
    <!-- <param name="detection_range" value="20"/> -->
    <!-- 地图输入方式 -->
    <!-- 0: 从world文件读取地图 -->
    <!-- 1: 从yaml文件读取地图 -->
    <!-- 2: 随机障碍物 -->
    <!-- 3: 读取点云信息 -->
    <param name="Map_input_method" value="2"/>
    <!-- gridmap 格子大小 -->
    <param name="gridmap_interval" value="0.025"/>
    <!-- 从world文件读取地图参数 -->
    <!-- 地图文件名称 -->
    <param name="world_file" value="$(find model_gazebo)/world/35boxes_10x10.world"/>
    <!-- 从yaml文件读取地图参数 -->
    <!-- 地图文件名称 -->
    <rosparam file="$(find model_gazebo)/config/map.yaml" command="load"/>
    <!-- 从ROS消息中接收点云 -->
    <!-- 地图文件名称 -->
    <param name="ROSpcl/map_range/x_min" value="-5"/>
    <param name="ROSpcl/map_range/x_max" value="5"/>
    <param name="ROSpcl/map_range/y_min" value="-5"/>
    <param name="ROSpcl/map_range/y_max" value="5"/>
    <!-- 随机障碍物 -->
    <!-- 地图文件名称 -->
    <param name="Random/map_range/x_min" value="-7"/>
    <param name="Random/map_range/x_max" value="7"/>
    <param name="Random/map_range/y_min" value="-7"/>
    <param name="Random/map_range/y_max" value="7"/>
    <!-- 圆形障碍物 没写-->
    <param name="Random/obstacle_sphere/num" value="1"/>
    <param name="Random/obstacle_sphere/radius" value="1"/>
    <!-- 正方形障碍物 -->
    <param name="Random/obstacle_box/num" value="75"/>
    <param name="Random/obstacle_box/length" value="1.0"/>
  </node>

  <node name="rviz" pkg="rviz" type="rviz" args="-d $(find model_gazebo)/rviz/octomap.rviz" />

  <node name="simulator" pkg="simulator" type="simulator" output="screen">
    <rosparam file="$(find plan_manager)/config/car.yaml" command="load"/>
    <param name="start_x" value="$(arg start_x_)"/>
    <param name="start_y" value="$(arg start_y_)"/>
    <param name="start_yaw" value="$(arg start_yaw_)"/>
    <param name="State_Propa_rate" value="500"/>
    <param name="Pose_pub_rate" value="100"/>
  </node>

  <node name="rqt_plot" pkg="rqt_plot" type="rqt_plot" args="/simulation/PosePub/v /simulation/PosePub/omega /simulation/PosePub/a /simulation/PosePub/alpha" />
  <!-- <node name="plotjuggler" pkg="plotjuggler" type="plotjuggler" output="screen" args="-l $(find simulator)/plotjuggler/plotjuggler.xml"/> -->
  <!-- <node name="plotjuggler" pkg="plotjuggler" type="plotjuggler" output="screen" args="-l $(find simulator)/plotjuggler/MPCDebug.xml"/> -->

  <group if="$(arg if_mpc_)">
    <arg name="odom_topic" value="/simulation/PosePub" />
    <arg name="traj_topic" value="/planner/traj_poly" />
    <arg name="cmd_topic" value="/simulation/PoseSub" />
    <node name="mpc" pkg="mpc_controller" type="mpc" output="screen">
      <rosparam file="$(find mpc_controller)/config/mpc.yaml" command="load"/>
      <remap from="cmd" to="$(arg cmd_topic)"/>
      <remap from="odom" to="$(arg odom_topic)"/>
      <remap from="traj" to="$(arg traj_topic)"/>
    </node>
  </group>
    
</launch>
